crates/exex/src/manager.rs

```diff
@@ -0,0 +1,466 @@
+
+            handle_rx,
+
+            min_id: 0,
+            next_id: 0,
+            buffer: VecDeque::with_capacity(max_capacity),
+            max_capacity,
+            current_capacity: Arc::clone(&current_capacity),
+
+            is_ready: is_ready_tx,
+            finished_height: finished_height_tx,
+
+            handle: ExExManagerHandle {
+                exex_tx: handle_tx,
+                num_exexs,
+                is_ready_receiver: is_ready_rx.clone(),
+                is_ready: WatchStream::new(is_ready_rx),
+                current_capacity,
+                finished_height: finished_height_rx,
+            },
+            metrics,
+        }
+    }
+
+    /// Returns the handle to the manager.
+    pub fn handle(&self) -> ExExManagerHandle {
+        self.handle.clone()
+    }
+
+    /// Updates the current buffer capacity and notifies all `is_ready` watchers of the manager's
+    /// readiness to receive notifications.
+    fn update_capacity(&mut self) {
+        let capacity = self.max_capacity.saturating_sub(self.buffer.len());
+        self.current_capacity.store(capacity, Ordering::Relaxed);
+        self.metrics.current_capacity.set(capacity as f64);
+        self.metrics.buffer_size.set(self.buffer.len() as f64);
+
+        // we can safely ignore if the channel is closed, since the manager always holds it open
+        // internally
+        let _ = self.is_ready.send(capacity > 0);
+    }
+
+    /// Pushes a new notification into the managers internal buffer, assigning the notification a
+    /// unique ID.
+    fn push_notification(&mut self, notification: CanonStateNotification) {
+        let next_id = self.next_id;
+        self.buffer.push_back((next_id, notification));
+        self.next_id += 1;
+    }
+}
+
+impl Future for ExExManager {
+    type Output = eyre::Result<()>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // drain handle notifications
+        while self.buffer.len() < self.max_capacity {
+            if let Poll::Ready(Some(notification)) = self.handle_rx.poll_recv(cx) {
+                debug!("received new notification");
+                self.push_notification(notification);
+                continue
+            }
+            break
+        }
+
+        // update capacity
+        self.update_capacity();
+
+        // advance all poll senders
+        let mut min_id = usize::MAX;
+        for idx in (0..self.exex_handles.len()).rev() {
+            let mut exex = self.exex_handles.swap_remove(idx);
+
+            // it is a logic error for this to ever underflow since the manager manages the
+            // notification IDs
+            let notification_id = exex
+                .next_notification_id
+                .checked_sub(self.min_id)
+                .expect("exex expected notification ID outside the manager's range");
+            if let Some(notification) = self.buffer.get(notification_id) {
```

Comment:
this lookup is a bit weird because this now looks up by index, I guess this is fine since we're keeping track of all the ids in the exexhandle, but ideally we lookup the buffer by id